atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCqlClusterContainer.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+import com.palantir.atlasdb.cassandra.CassandraServersConfigs.CassandraServersConfig;
+import com.palantir.atlasdb.cassandra.backup.CqlCluster;
+import com.palantir.atlasdb.keyvalue.cassandra.async.client.creation.ClusterFactory.CassandraClusterConfig;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.refreshable.Refreshable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * This container creates new CQL Clusters when the underlying server refreshable changes.
+ * After a refresh, previous CQL Clusters are closed to avoid resource leaks.
+ *
+ * There is no guarantee that a cluster provided via {@link #get} will <i>not</i> be closed whilst in use.
+ * Consumers should be resilient to any uses after the cluster has been closed, and where necessary, retry.
+ *
+ * After {@link #close()} has returned, no further CQL Clusters will be created, and all managed CQL Clusters will be
+ * closed.
+ */
+public final class ReloadingCqlClusterContainer implements Closeable, Supplier<CqlCluster> {
+    private static final SafeLogger log = SafeLoggerFactory.get(ReloadingCqlClusterContainer.class);
+
+    private final AtomicReference<Optional<CqlCluster>> lastCqlCluster;
```

Comment:
isn't this actually the "current" cql cluster, or something like that? i.e. at a snapshot in time, we expect this to match the refreshable, right?

Reply:
Heh I had latest and it somehow morphed to last. I'll call it current or latest.